File: mergeinterval/meetingrooms/minmeetingrooms/twopointer/MinMeetingRooms.java
package meetingrooms.minmeetingrooms.twopointer;

import java.util.Arrays;

class MinMeetingRooms {
    public int minMeetingRooms(int[][] meetings) {
        int totalMeetings = meetings.length;
        // Step 1: Extract start and end times of meetings
        int[] startTimes = new int[totalMeetings];
        int[] endTimes = new int[totalMeetings];

        for (int i = 0; i < totalMeetings; i++) {
            startTimes[i] = meetings[i][0];
            endTimes[i] = meetings[i][1];
        }

        // Step 2: Sort both start and end times
        Arrays.sort(startTimes);
        Arrays.sort(endTimes);

        // Step 3: Initialize two pointers for start and end times
        int currentRooms = 0;
        int maxRoomsNeeded = 0;
        int startPointer = 0;
        int endPointer = 0;

        // Step 4: Traverse through all meetings using startPointer
        while (startPointer < totalMeetings) {
            // If there's a meeting that has started before the previous one ends, allocate a new room
            if (startTimes[startPointer] < endTimes[endPointer]) {
                currentRooms++;
                startPointer++;
            } else {
                // Otherwise, a meeting has ended and we can free a room
                currentRooms--;
                endPointer++;
            }

            // Keep track of the maximum number of rooms required at any point
            maxRoomsNeeded = Math.max(maxRoomsNeeded, currentRooms);
        }

        return maxRoomsNeeded;
    }

    public static void main(String[] args) {
        MinMeetingRooms solution = new MinMeetingRooms();
        System.out.println(solution.minMeetingRooms(new int[][]{{0,30},{5,10},{15,20}}));  // 2
        System.out.println(solution.minMeetingRooms(new int[][]{{7,10},{2,4}}));  // 1
    }
}

File: mergeinterval/meetingrooms/minmeetingrooms/twopointer/MinMeetingRooms.py
class MinMeetingRooms:
    def __init__(self, intervals):
        self.intervals = intervals

    def minMeetingRooms(self):
        if not self.intervals:
            return 0

        # Step 1: Separate start and end times
        start_times = sorted([interval[0] for interval in self.intervals])
        end_times = sorted([interval[1] for interval in self.intervals])

        # Step 2: Use two pointers to count rooms
        start_pointer = 0
        end_pointer = 0
        rooms_required = 0
        max_rooms = 0

        while start_pointer < len(self.intervals):
            if start_times[start_pointer] < end_times[end_pointer]:
                # A new meeting is starting before the last one ends, need a new room
                rooms_required += 1
                start_pointer += 1
            else:
                # A meeting has ended, free up a room
                rooms_required -= 1
                end_pointer += 1

            # Track the maximum number of rooms required
            max_rooms = max(max_rooms, rooms_required)

        return max_rooms

# Example usage:
meeting_rooms_solver = MinMeetingRooms([[0, 30], [5, 10], [15, 20]])
print(meeting_rooms_solver.minMeetingRooms())  # Output: 2

meeting_rooms_solver_2 = MinMeetingRooms([[7, 10], [2, 4]])
print(meeting_rooms_solver_2.minMeetingRooms())  # Output: 1
